overwrite base config attribute

gleibovich-nvidia opened the pull request:
# What does this PR do ?

Creates a mechanism for resolving collisions between attributes found both in checkpoint and training configurations. Addresses the issue with the mechanism suggested in https://github.com/NVIDIA/NeMo-Aligner/issues/53.

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing a new algorithm
- [ ] Does the trainer resume and restore model state all states?
- [ ] Does the trainer support all parallelism techniques(PP, TP, DP)?
- [ ] Does the trainer support `max_steps=-1` and `validation`?
- [ ] Does the trainer only call APIs defined in [alignable_interface.py](/nemo_aligner/models/alignable_interface.py)?
- [ ] Does the trainer have proper logging?

# Additional Information
* Related to # (issue)

odelalleau commented on nemo_aligner/utils/utils.py:
```diff
@@ -123,7 +123,34 @@ def load_and_override_model_config(restore_path, model_cfg_to_overwrite, remove_
+
+            if isinstance(value, DictConfig) and isinstance(base[key], DictConfig):
+                recursive_merge(base[key], value, path=current_path)
+            else:
```
We should check if `value` is True (one reason to make it a config is so we can easily set it to False if needed)

odelalleau replied:
Resolving as this is addressed in my suggested commit

odelalleau commented on nemo_aligner/utils/utils.py:
```diff
@@ -123,7 +123,34 @@ def load_and_override_model_config(restore_path, model_cfg_to_overwrite, remove_
+                current_config = overwrite_base_config
+                for sub_key in current_path:
+                    if isinstance(current_config, DictConfig):
+                        current_config = current_config[sub_key] if sub_key in current_config else None
```
I'm a bit confused by what this `current_config` is doing. Since we're going down the same path we went through aren't we guaranteed that `overwrite_base_config` has such a path? And isn't `current_config` essentially the same as `value`?

odelalleau replied:
Resolving this comment as I had misunderstood this code.

odelalleau commented on nemo_aligner/utils/utils.py:
```diff
@@ -123,7 +123,34 @@ def load_and_override_model_config(restore_path, model_cfg_to_overwrite, remove_
+
+
+def resolve_overlapping_fields(base_config, overwrite_config):
+    def recursive_merge(base, overwrite, path=[]):
```
Just to be safe, better avoid a mutable default (using `None` instead)

odelalleau commented on nemo_aligner/utils/utils.py:
```diff
@@ -123,7 +123,34 @@ def load_and_override_model_config(restore_path, model_cfg_to_overwrite, remove_
+    overwrite_base_config = overwrite_config.get("overwrite_base_config", OmegaConf.create())
+    recursive_merge(base_config, overwrite_config)
+
+    return DictConfig(base_config)
```
Why call `DictConfig` here? We've already mutated the input base config (may be safer not to => deep copy it first instead?)

odelalleau reviewed (commented):
The high-level logic seems fine but I find the main function doing the resolution a bit hard to understand. I feel like it can probably be made simpler.

gleibovich-nvidia commented:
> The high-level logic seems fine but I find the main function doing the resolution a bit hard to understand. I feel like it can probably be made simpler.

Yeah I agree it's confusing. I think it's mainly because of limited APIs from `OmegaConf.DictConfig`. Does addressing the comments raised made it clearer now? Did you had some other logic in mind which would have made it simpler?

odelalleau reviewed (commented):
> Does addressing the comments raised made it clearer now? Did you had some other logic in mind which would have made it simpler?

It was a bit difficult for me to wrap my head around this yesterday without actually running the code. After running it I came up with 9487d754 which you can cherry-pick on top of your branch if you like it.

This commit:
* Makes things simpler overall (IMO) -- no more nested recursive function dealing with 3 configs, we just walk down `overwrite_base_config` along with the base config to identify which keys should be removed
* Fixes a couple of issues: (1) the value of the flag (True or False) used to be ignored and (2) the keys being popped were the children of the key we actually want to pop, which I believe wouldn't have worked with a `ListConfig`
* Actually modifies `base_config` in place: I know I suggested not to do this in my previous review, but I hadn't realized that this base config  was created from within `load_and_override_model_config()` and thus it seems safe to me (while keeping the code simpler)
* Removes the `overwrite_base_config` key from the final model config (this is just to avoid cluttering the config -- we don't need this key anymore once the config is merged)

Let me know what you think!

odelalleau commented on examples/nlp/gpt/conf/gpt_dpo.yaml:
```diff
@@ -97,4 +97,8 @@ model:
     index_mapping_dir: null # path to save index mapping .npy files, by default will save in the same location as data_prefix
     data_prefix: null
 
+  overwrite_base_config:
```
Let's add a comment to make it more understandable (please do it in all files -- feel free to rewrite):

```suggestion
  # Which fields from the base model config should be ignored when merging with this config.
  overwrite_base_config:
```

odelalleau commented on examples/nlp/gpt/conf/inference_rm.yaml:
```diff
@@ -24,4 +24,8 @@ model:
\ No newline at end of file
+    attribute_weights: null # apply these weights to each attributes when merging them into a scalar
+
+  overwrite_base_config:
```
Can be removed from this file (it doesn't have a `data_prefix`)

odelalleau commented on examples/nlp/gpt/conf/gpt_ppo_critic.yaml:
```diff
@@ -96,4 +96,8 @@ model:
       constant_steps: 1000
       min_lr: 9e-7
 
+  overwrite_base_config:
```
Can be removed from this file (it doesn't have a `data_prefix`)

odelalleau commented on examples/nlp/gpt/conf/gpt_sft.yaml:
```diff
@@ -165,6 +165,10 @@ model:
       constant_steps: 1000
       min_lr: 9e-7
 
+  overwrite_base_config:
```
Can be removed from this file (it doesn't have a `data_prefix`)

odelalleau reviewed (changes requested):
Looks good, just some final suggestions before merging.

In addition to the comments below, please also add a CHANGELOG entry (it's a bugfix, something like "fix crash that could occur when the base model's `data.data_prefix` config was a list")

gleibovich-nvidia commented:
> Looks good, just some final suggestions before merging.
> 
> In addition to the comments below, please also add a CHANGELOG entry (it's a bugfix, something like "fix crash that could occur when the base model's `data.data_prefix` config was a list")

Sure, done.

odelalleau commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 ### Breaking changes
 
 ### Bug Fixes
+- Fixed a potential issue when the base model's `model.data.data_prefix` config is a list and is about to overriden with
```
```suggestion
- Fixed a potential issue when the base model's `model.data.data_prefix` config is a list and is about to be overridden with
```

odelalleau reviewed (dismissed):
Please just fix the minor typo in changelog then good to merge!

odelalleau reviewed (dismissed):
Sorry I didn't realize I needed to re-review, all good and thanks a lot for the help!

gleibovich-nvidia commented:
> Sorry I didn't realize I needed to re-review, all good and thanks a lot for the help!

Merged `main` into the PR, and now need to re-review (again) in order to merge :-)

odelalleau reviewed (dismissed):
haha :D

odelalleau reviewed (approved):
Fun times!
